### PR TITLE
Add C# 3LO and API key auth support

### PR DESCRIPTION
**Pull request by saicheems**

### CODE REVIEW

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_adexchangebuyer.v1.4.json.baseline [hunk 1]
```diff
@@ -54,6 +42,12 @@ namespace AdExchangeBuyerSample
             // To execute asynchronously in an async method, use
             // Data.Account response = await request.ExecuteAsync();
         }
+
+        public static UserCredential getCredential() {
+            // TODO: Implement this function to get authentication credentials.
+            // See https://foo.com/bar
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_adexchangebuyer.v1.4.json.baseline]: The scopes should be showing up here.

File: src/main/resources/com/google/api/codegen/csharp/discovery_fragment.snip [hunk 2]
```diff
@@ -44,31 +44,66 @@
...
+            @end
+        @end
+        UserCredential credential = GetCredential();
+        {@sample.serviceTypeName} {@sample.serviceVarName} = new {@sample.serviceTypeName}(new BaseClientService.Initializer
+        {
+            HttpClientInitializer = credential,
+            ApplicationName = "Google Cloud Platform Sample",
+        });
+    @default
+        // TODO: Replace "{API-KEY}"
```
Comment (vchudnov-g) [inline src/main/resources/com/google/api/codegen/csharp/discovery_fragment.snip]: Do we need a `{MY-` prefix?

File: src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_adexchangebuyer.v1.4.json.baseline (hunk 1 above)
Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_adexchangebuyer.v1.4.json.baseline]: Fixed

Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_adexchangebuyer.v1.4.json.baseline]: The boilerplate at the top of this snippet is still referring to ADC.

Comment (saicheems) [inline src/test/java/com/google/api/codegen/testdata/discoveries/csharp/csharp_adexchangebuyer.v1.4.json.baseline]: Fixed (nice catch)

**Comment by vchudnov-g** [comment]
LGTM

**Comment by saicheems** [comment]
Compilecheck works for the grand majority of APIs, where 2 APIs had minor errors related that must be fixed with overrides (type name mismatches).

**Pull request merged**